From 4854b944b345990e4100319662777856fe8ea4aa Mon Sep 17 00:00:00 2001
From: Adrian Perez de Castro <aperez@igalia.com>
Date: Thu, 15 Aug 2024 13:15:42 -0700
Subject: [PATCH] Cherry-pick 282306@main (96fb0b0c6c46).
 https://bugs.webkit.org/show_bug.cgi?id=274780

    [WPE][GTK] Crash in WebCore::TextDecorationPainter::paintBackgroundDecorations when compiled with Clang with LTO enabled
    https://bugs.webkit.org/show_bug.cgi?id=274780

    Reviewed by Michael Catanzaro.

    Clang seem to have some issue when inlining assignment and move
    operators in LTO builds, generating code that tries to perform OOB
    access to Vector data. Replacing an assignmenmt with a Vector::swap(),
    which is semantically equivalent in this case (the moved-from object
    is not used again in the function) workarounds the compiler issue.

    * Source/WebCore/rendering/TextDecorationPainter.cpp:
    (WebCore::translateIntersectionPointsToSkipInkBoundaries):

    Canonical link: https://commits.webkit.org/282306@main

Canonical link: https://commits.webkit.org/274313.374@webkitglib/2.44
---
 Source/WebCore/rendering/TextDecorationPainter.cpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/Source/WebCore/rendering/TextDecorationPainter.cpp b/Source/WebCore/rendering/TextDecorationPainter.cpp
index 229857f64ad4..a58bbd86f452 100644
--- a/Source/WebCore/rendering/TextDecorationPainter.cpp
+++ b/Source/WebCore/rendering/TextDecorationPainter.cpp
@@ -140,8 +140,10 @@ static DashArray translateIntersectionPointsToSkipInkBoundaries(const DashArray&
             else
                 intermediateTuples.append(*i);
         }
-    } else
-        intermediateTuples = tuples;
+    } else {
+        // XXX(274780): A plain assignment or move here makes Clang generate bad code in LTO builds.
+        intermediateTuples.swap(tuples);
+    }
 
     // Step 3: Output the space between the ranges, but only if the space warrants an underline.
     float previous = 0;
